derive/src/to_delegate.rs

```diff
@@ -206,21 +206,39 @@ impl RpcMethod {
 	}
 
 	fn generate_delegate_closure(&self, is_subscribe: bool) -> Result<proc_macro2::TokenStream> {
-		let mut param_types: Vec<_> = self
+		let args = self
 			.trait_item
 			.sig
 			.inputs
 			.iter()
 			.cloned()
-			.filter_map(|arg| match arg {
-				syn::FnArg::Typed(ty) => Some(*ty.ty),
-				_ => None,
+			.filter_map(|arg| {
+				if let syn::FnArg::Typed(pat_type) = arg {
+					Some(pat_type)
+				} else {
+					None
+				}
 			})
-			.collect();
+			.enumerate();
 
 		// special args are those which are not passed directly via rpc params: metadata, subscriber
-		let special_args = Self::special_args(&param_types);
-		param_types.retain(|ty| !special_args.iter().any(|(_, sty)| sty == ty));
+		let mut special_args = vec![];
+		let mut fn_args = vec![];
+
+		for (i, arg) in args {
+			if let Some(sarg) = Self::special_arg(i, arg.clone()) {
+				special_args.push(sarg);
+			} else {
+				fn_args.push(arg);
+			}
+		}
+
+		let special_args: Vec<_> = special_args;
+		let fn_args: Vec<_> = fn_args;
+
+		let param_types: Vec<_> = fn_args.iter().cloned().map(|arg| *arg.ty).collect();
+		let arg_names: Vec<_> = fn_args.iter().cloned().map(|arg| *arg.pat).collect();
```

Comment:
```suggestion
		let param_types = fn_args.iter().map(|arg| *arg.ty).collect::<Vec<_>>();
		let arg_names = fn_args.iter().map(|arg| *arg.pat).collect::<Vec<_>>();
```
My preference would be to add a generic parameter to `collect`, but I'm not super strong on this.

Also `.cloned()` doesn't seem required if we dereference/copy the type, is it?

Reply:
`syn::Type` and `syn::Pat` don't implement `Copy`, so there must be a `.clone()` somewhere.
It should be more efficient to clone only `syn::Type` or `syn::Pat` instead of the whole `syn::FnArg`.